comp/core/secrets/secretsimpl/secrets_test.go

```diff
@@ -366,30 +367,105 @@ func TestResolveWithCallback(t *testing.T) {
 	topLevelResolved := 0
 	secondLevelResolved := 0
 	thirdLevelResolved := 0
-	err := resolver.ResolveWithCallback(
-		testConf,
-		"test",
-		func(yamlPath []string, value any) bool {
-			switch strings.Join(yamlPath, "/") {
-			case "top_level":
-				assert.Equal(t, "password1", value)
-				topLevelResolved++
-			case "some/second_level":
-				assert.Equal(t, "password2", value)
-				secondLevelResolved++
-			case "some/encoded/third_level":
-				assert.Equal(t, "password3", value)
-				thirdLevelResolved++
-			default:
-				assert.Fail(t, "unknown yaml path: %s", yamlPath)
-			}
-			return true
-		},
-	)
+	resolver.SubscribeToChanges(func(handle string, path []string, oldValue, newValue any) {
+		switch strings.Join(path, "/") {
+		case "top_level":
+			assert.Equal(t, "password1", newValue)
+			topLevelResolved++
+		case "some/second_level":
+			assert.Equal(t, "password2", newValue)
+			secondLevelResolved++
+		case "some/encoded/third_level":
+			assert.Equal(t, "password3", newValue)
+			thirdLevelResolved++
+		default:
+			assert.Fail(t, "unknown yaml path: %s", path)
+		}
+	})
+	_, err := resolver.Resolve(testConf, "test")
+
 	require.NoError(t, err)
 	assert.Equal(t, 1, topLevelResolved, "'top_level' secret was not resolved or resolved multiple times")
 	assert.Equal(t, 1, secondLevelResolved, "'second_level' secret was not resolved or resolved multiple times")
-	assert.Equal(t, 1, thirdLevelResolved, "'third_level' secret was not resolved or resolved multiple times")
+	assert.Equal(t, 0, thirdLevelResolved, "'third_level' secret was resolved, but should not have been because it was already in the cache")
 
 	assert.Equal(t, testConfNestedOriginMultiple, resolver.origin)
 }
+
+func TestResolveThenRefresh(t *testing.T) {
+	testConf := testConfNestedMultiple
+
+	// disable the whitelist for the test, let any secret changes happen
+	originalWhitelistHandles := whitelistHandles
+	whitelistHandles = nil
+	defer func() { whitelistHandles = originalWhitelistHandles }()
```

Comment:
I'm not sure if you added it already, but should we add a test that exercises `whitelistHandles` having a value and only subscription being called for that value and not others?